### PR TITLE
#3667 - Bull Scheduler Increments the Next Scheduled Job When Manually Promoted

### PR DESCRIPTION
**Pull request by andrewsignori-aot**
Updated Bull Board-related packages to take advantage of the feature present in the newer version that allows a new job to be added (same as duplicate), allowing the schedulers to be executed without affecting their current time to be executed again.

_Notes:_
1. during the local tests, when a delayed job was deleted and the `queue-consumers` restarted, the delayed job was restored.
2. this PR is intended to be part of the upcoming release. The research for the ticket will continue.

## Using the add/duplicate option

While creating the new job the properties can be edited, for instance, the corn expression can be edited to create a new time to execute the scheduler.

_Please note that `timestamp` and `prevMillis` should be removed. These properties are generated once the job is created.
As mentioned in the source code, `timestamp` is the "Timestamp when the job was created." and  `prevMillis` is a "Internal property used by repeatable jobs."_

image

```json
{
  "repeat": {
    "count": 1,
    "key": "__default__::::0 7 * * *",
    "cron": "0 7 * * *"
  },
  "jobId": "repeat:2c2720c5e8b4e9ce99993becec27a0ff:1731999600000",
  "delay": 42485905,
  "timestamp": 1731957114095,
  "prevMillis": 1731999600000,
  "attempts": 3,
  "backoff": {
    "type": "fixed",
    "delay": 180000
  }
}
```

## Refactor

- Refactored the code to move the Bull Board configuration to the Nestjs modules instead of doing it on the main.ts.
The refactored code is equivalent to the one previously on the `main.ts`.
- Updated icons and labels to make the dashboard look more like part of the SIMS. The way it was done was by targeting less effort. In case it causes some noise during the PR review the code will be removed.

### CODE REVIEW

File: sources/packages/backend/libs/services/src/constants/system-configurations-constants.ts [hunk 1]
```diff
@@ -90,3 +90,8 @@ export const BLOCKED_DISBURSEMENT_NOTIFICATION_MIN_DAYS_INTERVAL = 7;
  * Represents the maximum number of disbursement blocked notifications to send.
  * */
 export const BLOCKED_DISBURSEMENT_MAXIMUM_NOTIFICATIONS_TO_SEND = 3;
+
+/**
+ * Bull Dashboard route.
+ */
+export const BULL_BOARD_ROUTE = "admin/queues";
```
Comment (dheepak-aot) [inline sources/packages/backend/libs/services/src/constants/system-configurations-constants.ts:97]: 👍

File: sources/packages/backend/libs/services/src/queue/queue.module.ts [hunk 2]
```diff
@@ -25,6 +30,12 @@ import { DatabaseModule } from "@sims/sims-db";
       inject: [ConfigService],
     }),
     BullModule.registerQueueAsync(...getQueueModules()),
+    BullBoardModule.forRootAsync({
```
Comment (dheepak-aot) [inline sources/packages/backend/libs/services/src/queue/queue.module.ts:33]: As discussed, the queue module is shared with API (to publish messages to queue) and loading bull board module here will stand up the bull dashboard in SIMS API as well.

File: sources/packages/backend/libs/services/src/queue/queue.module.ts [hunk 3]
```diff
@@ -91,3 +102,36 @@ function getQueueModules(): BullModuleAsyncOptions[] {
...
   }));
 }
+
+/**
+ * Builds the Bull Board module options to register the dashboard in a dynamic way.
+ * @param configService service with the configuration of the application.
+ * @returns Bull Board module options with the dashboard route,
+ * authentication middleware and the board options.
+ */
+async function bullBoardModuleFactory(configService: ConfigService) {
```
Comment (dheepak-aot) [inline sources/packages/backend/libs/services/src/queue/queue.module.ts:112]: Please specify the return type in the method.

```ts
async function bullBoardModuleFactory(
  configService: ConfigService,
): Promise<BullBoardModuleOptions>
```

File: sources/packages/backend/libs/services/src/queue/bull-board-queues-module.ts [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+    private readonly moduleRef: ModuleRef,
+    private readonly queueService: QueueService,
+    @Inject(BULL_BOARD_INSTANCE)
+    private readonly board: BullBoardInstance,
+  ) {}
+
+  /**
+   * Adds all queues to the bull board during application initialization.
+   * checking if the queue is active and if it is a scheduler.
+   */
```
Comment (dheepak-aot) [inline sources/packages/backend/libs/services/src/queue/bull-board-queues-module.ts:22]: Minor please check the comment and period.

Also please add returns.

File: sources/packages/backend/libs/services/src/queue/queue.module.ts [hunk 5]
```diff
@@ -91,3 +102,36 @@ function getQueueModules(): BullModuleAsyncOptions[] {
...
+    middleware: authMiddleware,
+    boardOptions: {
+      uiConfig: {
+        boardTitle: "SIMS-Queues",
+        boardLogo: {
+          path: "https://sims.studentaidbc.ca/favicon-32x32.png",
+        },
+        favIcon: {
+          default: "https://sims.studentaidbc.ca/favicon-16x16.png",
+          alternative: "https://sims.studentaidbc.ca/favicon-32x32.png",
```
Comment (dheepak-aot) [inline sources/packages/backend/libs/services/src/queue/queue.module.ts:132]: ❤️Looks great with SIMS logo.

**Comment by dheepak-aot** [comment]
Great Job! Please have a look at the comments. Except one rest all are minor.

File: sources/packages/backend/libs/services/src/queue/queue.module.ts (hunk 2 above)
Comment (andrewsignori-aot) [reply sources/packages/backend/libs/services/src/queue/queue.module.ts]: The modules are now added only to the queue-consumers.

File: sources/packages/backend/libs/services/src/queue/queue.module.ts (hunk 3 above)
Comment (andrewsignori-aot) [reply sources/packages/backend/libs/services/src/queue/queue.module.ts]: Return type added.

File: sources/packages/backend/libs/services/src/queue/bull-board-queues-module.ts (hunk 4 above)
Comment (andrewsignori-aot) [reply sources/packages/backend/libs/services/src/queue/bull-board-queues-module.ts]: Add the period.
About the return type, I do not believe that we usually add comments for Promise<void>. Please let me know if that is what you meant.

**Review by guru-aot — APPROVED** [review]
LGTM, nice work @andrewsignori-aot

File: sources/packages/backend/libs/services/src/queue/bull-board-queues-module.ts (hunk 4 above)
Comment (dheepak-aot) [reply sources/packages/backend/libs/services/src/queue/bull-board-queues-module.ts]: You are right about the return type.

File: sources/packages/backend/apps/queue-consumers/test/helpers/testing-modules/testing-modules-helper.ts [hunk 6]
```diff
@@ -45,6 +47,10 @@ export async function createTestingAppModule(): Promise<CreateTestingModuleResul
       replace: QueueModule,
       by: QueueModuleMock,
     },
+    {
+      replace: BullBoardQueuesModule,
+      by: BullBoardQueuesModuleMock,
```
Comment (dheepak-aot) [inline sources/packages/backend/apps/queue-consumers/test/helpers/testing-modules/testing-modules-helper.ts:52]: 👍

**Review by dheepak-aot — APPROVED** [review]
Thanks for doing the changes. Looks good 👍

**Pull request merged**